cEP-0032.md

```diff
@@ -0,0 +1,651 @@
+                  # 3. Valid newcomer issue(link provided)(optional)
+                  # 4. Get user name and bio from github
+                  # 5. set teams to coala_newcomers
+                  # 6. Check whether is already a member or not
+                  # 7. set status of account to disable
+                  # 8. Perform mentioned checks too in issue
+                  return data
+
+### Want to get promoted to developer? A.k.a Promotion to Developer
+
+* coala has a reward or a surprise gift for newcomers who have completed the
+newcomers process as mentioned in Newcomer’s guide.
+* This process can be automated by performing some rudimentary checks to
+promote the newcomer to developer.
+* A newcomer can choose “Promote to Developer” from profile page on community
+website. The user can apply for this at most 3 times.
+* API Endpoint:
+
+            from .views import NewcomerPromotion
+            urlpatterns = [
+            path('^newcomer/promote/', NewcomerPromotion.as_view(), name='Newcomer promotion')
+            ]
+
+* Promotion class view:
+
+            from rest_framework.generics import UpdateAPIView
+            from .models import Contributor
+            from .serializers import NewcomerPromoteSerializer
+
+            class NewcomerPromotion(UpdateAPIView):
+            permission_classes = []
+            authentication_classes = []
+            serializer_class = NewcomerPromoteSerializer
+            queryset = Contributor.objects.all()
+
+* Respective Serializer class for validating:
+
+            from rest_framework import serializers
+            from .models import Contributor
+
+            class NewcomerPromoteSerializer(serializers.ModelSerializer):
+            class Meta:
+                  model = Contributor
+                  fields = '__all__'
+
+            def validate(self, data):
+                  # Validate POST data
+                  # Evaluate the google form
+                  # 1. Get user latest entry from response spreadsheet
+                  # 2. Valid data i.e.
+                  #        - username
+                  #        - solved newcomer issue
+                  #        - solver low difficult issue
+                  #        - reviewed two PRs
+                  return data
+
+* Validating this data will definitely require some amount of time.
+  This can be achieved in two ways:
+  1. One way is that user chooses to get promoted. Then, the API call will be
+  made in background and user can’t do any work until the validating process is
+  completed.
+  2. Other way is create a database model which will have usernames with other
+  information. And, then a cron Job will be executed for all such usernames to
+  validate their information. The job will be executed for only those username
+  who job scheduled was pending.
+
+      I would like to go for second one during coding if I get enough time to
+      achieve this.
+
+### Authentication using GitHub
```

Comment:
I don't think auth should be done this way. `django-allauth` has other methods of login that support other platforms more suitable than GitHub. Sould look into any other OAuth provider.

Reply:
Can you help me out by suggesting other OAuth providers that can help us to achieve this task?

Reply:
Maybe if we leave the new endpoints as a pub-sub service rather than requiring auth when there are forms. I think that rather than finding other providers, it is more suitable to focus at one approach. Either go for form-based and then assume if completion is necessary to create coala auth grants or for GitHub. I think the `django-allauth` approach is ok but make sure that both GitHub and GitLab are implemented.